File: backend/src/controllers/finance.controller.ts
import { Request, Response } from 'express'
import { supabase } from '../config/supabase'

export async function listarTransacoes(req: Request, res: Response) {
  const userId = (req as any).user.id
  const { data, error } = await supabase
    .from('transactions')
    .select('*')
    .eq('user_id', userId)
    .order('date', { ascending: false })
    .order('created_at', { ascending: false })
  if (error) return res.status(500).json({ error: error.message })
  return res.json(data)
}

export async function criarTransacao(req: Request, res: Response) {
  const userId = (req as any).user.id
  const { description, amount, type, category, label, date, recurring } = req.body
  if (!description?.trim()) return res.status(400).json({ error: 'Descrição obrigatória' })
  if (!amount || amount <= 0)  return res.status(400).json({ error: 'Valor inválido' })
  if (!['income','expense'].includes(type)) return res.status(400).json({ error: 'Tipo inválido' })

  const { data, error } = await supabase
    .from('transactions')
    .insert({ user_id: userId, description: description.trim(), amount, type, category, label: label ?? '', date, recurring: recurring ?? false })
    .select()
    .single()
  if (error) return res.status(500).json({ error: error.message })
  return res.status(201).json(data)
}

export async function editarTransacao(req: Request, res: Response) {
  const userId = (req as any).user.id
  const { id } = req.params
  const { description, amount, type, category, label, date, recurring } = req.body

  if (amount !== undefined && amount <= 0) return res.status(400).json({ error: 'Valor inválido' })
  if (type !== undefined && !['income','expense'].includes(type)) return res.status(400).json({ error: 'Tipo inválido' })

  const updates: Record<string, unknown> = {}
  if (description !== undefined) updates.description = description.trim()
  if (amount      !== undefined) updates.amount      = amount
  if (type        !== undefined) updates.type        = type
  if (category    !== undefined) updates.category    = category
  if (label       !== undefined) updates.label       = label
  if (date        !== undefined) updates.date        = date
  if (recurring   !== undefined) updates.recurring   = recurring

  const { data, error } = await supabase
    .from('transactions').update(updates).eq('id', id).eq('user_id', userId).select().single()
  if (error) return res.status(500).json({ error: error.message })
  return res.json(data)
}

export async function deletarTransacao(req: Request, res: Response) {
  const userId = (req as any).user.id
  const { id } = req.params
  const { error } = await supabase.from('transactions').delete().eq('id', id).eq('user_id', userId)
  if (error) return res.status(500).json({ error: error.message })
  return res.status(204).send()
}

File: backend/internal/finance/finance_controller.go
package finance

import (
	"net/http"
	"strconv"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"beeplanerofc/backend-go/internal/dbutil"
	"beeplanerofc/backend-go/internal/middleware"
)

type Controller struct {
	db *pgxpool.Pool
}

func NewController(db *pgxpool.Pool) *Controller {
	return &Controller{db: db}
}

var tiposValidos = map[string]bool{"income": true, "expense": true}

// @Summary      Listar transações
// @Tags         Finance
// @Produce      json
// @Success      200 {array} map[string]interface{}
// @Security     BearerAuth
// @Router       /finance [get]
func (ctrl *Controller) ListarTransacoes(c *gin.Context) {
	user := middleware.CurrentUser(c)

	rows, err := ctrl.db.Query(c.Request.Context(),
		"SELECT * FROM transactions WHERE user_id = $1 ORDER BY date DESC, created_at DESC", user.ID,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	transacoes, err := pgx.CollectRows(rows, dbutil.RowToMap)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, transacoes)
}

type criarTransacaoRequest struct {
	Description string   `json:"description"`
	Amount      *float64 `json:"amount"`
	Type        string   `json:"type"`
	Category    string   `json:"category"`
	Label       *string  `json:"label"`
	Date        string   `json:"date"`
	Recurring   *bool    `json:"recurring"`
}

// @Summary      Criar transação
// @Tags         Finance
// @Accept       json
// @Produce      json
// @Param        body body criarTransacaoRequest true "Dados da transação"
// @Success      201 {object} map[string]interface{}
// @Security     BearerAuth
// @Router       /finance [post]
func (ctrl *Controller) CriarTransacao(c *gin.Context) {
	user := middleware.CurrentUser(c)

	var req criarTransacaoRequest
	c.ShouldBindJSON(&req)

	description := strings.TrimSpace(req.Description)
	if description == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Descrição obrigatória"})
		return
	}
	if req.Amount == nil || *req.Amount <= 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Valor inválido"})
		return
	}
	if !tiposValidos[req.Type] {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Tipo inválido"})
		return
	}

	label := ""
	if req.Label != nil {
		label = *req.Label
	}
	recurring := false
	if req.Recurring != nil {
		recurring = *req.Recurring
	}

	rows, err := ctrl.db.Query(c.Request.Context(), `
		INSERT INTO transactions (user_id, description, amount, type, category, label, date, recurring)
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
		RETURNING *`,
		user.ID, description, *req.Amount, req.Type, req.Category, label, req.Date, recurring,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	transacao, err := pgx.CollectOneRow(rows, dbutil.RowToMap)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusCreated, transacao)
}

type editarTransacaoRequest struct {
	Description *string  `json:"description"`
	Amount      *float64 `json:"amount"`
	Type        *string  `json:"type"`
	Category    *string  `json:"category"`
	Label       *string  `json:"label"`
	Date        *string  `json:"date"`
	Recurring   *bool    `json:"recurring"`
}

// @Summary      Editar transação
// @Tags         Finance
// @Accept       json
// @Produce      json
// @Param        id   path string                true "ID da transação"
// @Param        body body editarTransacaoRequest true "Campos a atualizar"
// @Success      200 {object} map[string]interface{}
// @Failure      404 {object} map[string]string
// @Security     BearerAuth
// @Router       /finance/{id} [put]
func (ctrl *Controller) EditarTransacao(c *gin.Context) {
	user := middleware.CurrentUser(c)
	id := c.Param("id")

	var req editarTransacaoRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Dados inválidos"})
		return
	}

	if req.Amount != nil && *req.Amount <= 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Valor inválido"})
		return
	}
	if req.Type != nil && !tiposValidos[*req.Type] {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Tipo inválido"})
		return
	}

	setClauses := []string{}
	args := []any{}
	addField := func(column string, value any) {
		args = append(args, value)
		setClauses = append(setClauses, column+" = $"+strconv.Itoa(len(args)))
	}

	if req.Description != nil {
		addField("description", strings.TrimSpace(*req.Description))
	}
	if req.Amount != nil {
		addField("amount", *req.Amount)
	}
	if req.Type != nil {
		addField("type", *req.Type)
	}
	if req.Category != nil {
		addField("category", *req.Category)
	}
	if req.Label != nil {
		addField("label", *req.Label)
	}
	if req.Date != nil {
		addField("date", *req.Date)
	}
	if req.Recurring != nil {
		addField("recurring", *req.Recurring)
	}

	if len(setClauses) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Nenhum campo para atualizar"})
		return
	}

	args = append(args, id, user.ID)
	query := "UPDATE transactions SET " + strings.Join(setClauses, ", ") +
		" WHERE id = $" + strconv.Itoa(len(args)-1) + " AND user_id = $" + strconv.Itoa(len(args)) +
		" RETURNING *"

	rows, err := ctrl.db.Query(c.Request.Context(), query, args...)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	transacao, err := pgx.CollectOneRow(rows, dbutil.RowToMap)
	if err == pgx.ErrNoRows {
		c.JSON(http.StatusNotFound, gin.H{"error": "Transação não encontrada"})
		return
	}
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, transacao)
}

// @Summary      Deletar transação
// @Tags         Finance
// @Produce      json
// @Param        id path string true "ID da transação"
// @Success      204
// @Security     BearerAuth
// @Router       /finance/{id} [delete]
func (ctrl *Controller) DeletarTransacao(c *gin.Context) {
	user := middleware.CurrentUser(c)
	id := c.Param("id")

	_, err := ctrl.db.Exec(c.Request.Context(),
		"DELETE FROM transactions WHERE id = $1 AND user_id = $2", id, user.ID,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusNoContent)
}
